### ISSUE TITLE
Directly load / refresh worlds in Blender

### ISSUE THREAD

**Issue by TheDuckCow** [TheDuckCow/MCprep#75]
## Feature overview

Vision: To directly load in Minecraft saves without ever leaving Blender.

How? By using the scripting feature of Mineways, which is a way to control the application without user interactions. Calls to the program will still generate OBJ files which are then imported into blender automatically, seamlessly for the user.

Why not natively implement loading of worlds like e.g. MineBlend? Well, that's a ton of work, and hard to keep up to date, and as soon as this feature is created, users will depend on it heavily and it needs to be *stable*. Instead of reinventing an entire wheel, MCprep will lean on the existing and still maintained Mineways project. 

## Implementation steps

- [ ] Pure pythonic proof of concept controlling Mineways on both windows and mac
- [ ] Basic implementation inside of Blender (manually specified coordinates, few options)
- [ ] Splitting of import regions into chunks, with individuals 
- [ ] Workflow or tools to let user easily select _what_ region they want to import
- [ ] Reloading single chunk workflow
- [ ] Reloading multiple workflow, with height changes
- [ ] World extension workflow

## Implementation architecture

This will be implemented with a split between a pure-pythonic wrapper which calls/communicates with Mineways, and then a parent blender python file which implements the blender-specific operators and user workflow.

Currently, Mineways can only be launched with scripts as a starting input script file, and after this script executes, the program either is still open (ie "orphaned") or closes (if the close command is included in the script). Because we want there to be no user interaction with Mineways itself, every command needs to end with a close statement. 

To keep blender responsive, this should be implemented in a parallel process so that there can be interactions such as ESC to stop the current import where it is, or to cancel altogether. Because we cannot communicate to an already opened and running instance of Mineways (yet), this means batching the world out into chunks (where Mineways opens, exports a chunk or two, then completely closes down) and then offering to stop after some have ran, but always letting Mineways finish its current process. 

To be decided: have this all running in a modal operator (most stable), or allow the parallel process to continue while user can interact with blender as chunks are loaded (likely quite unstable). 

Additionally, the world should be managed and thought of more as a "bridge" than a once imported and done approach. Settings for how a world was imported will be saved, and features will allow for piece-wise reimporting of chunks, or even extending the imported world as-needed. 

## Links and references

- [Early teaser of the feature](https://twitter.com/TheDuckCow/status/1151276654787989505)
- [Reference for (stale) MineBlend](https://sites.google.com/site/mineblend/), which technically works still (2.7x) but not complete or comprehensive
  - Could get some reference code for bare minimum save data file reading, but avoid and leverage Mineways for everything possible.

**Comment by TheDuckCow** [comment]
Realized now that a command-line interface also already exists for jmc2obj, and should sufficiently work to include as an alternate option when using the bridge feature. 

Sample execution to output a world to obj (and exporting the texture folder):
```
java -jar ./jMc2Obj-107.jar \
  --dimension=0 --area=-40,-40,40,40 --height=0,256 \
  --render-entities --object-per-mat \
  --output=/..../test_jmc_cmd/ \
  --objfile=test_jmc2_cmdlin.obj --mtlfile=test_jmc2_cmdlin.mtl \
  --export=obj,mtl,tex "..../saves/world_save"
```

More options visible via --help. Note to self that I noticed at first the tex exporting was rather slow. Also calling out that newest jmc2obj appears to not be able to open old-format chunks, and so can only be used with newest world formats, which should generally be fine for users anyways, as they should be able ot just open their old worlds in the latest Minecraft version..

**Comment by TheDuckCow** [comment]
While little additional development done, I have much more of the design approach worked out. See the architecture diagram here (click for higher resolution):

![MCprep Connector Architecture](https://user-images.githubusercontent.com/2958461/63235105-38cc0d80-c206-11e9-89f5-9c27c74c122d.png)


This approach was designed keeping in mind submissions from [the survey](https://t.co/yfDr87BOMl?amp=1), where the biggest insights were:
- Yes, people do tend to have multiple worlds imported inside a blend file at once
- Yes, people tend to wish they could reload a world already imported
- Overwhelmingly yes, they wish they could extend a world after an initial import.

I was not strongly in favor of supporting multiple worlds linked at once, but I see now it's important enough to strive for.

## Source from survey (although only 12 responses):
![re-import](https://user-images.githubusercontent.com/2958461/63235472-8ac16300-c207-11e9-9a27-296b3526deee.png)



![multi world](https://user-images.githubusercontent.com/2958461/63235144-5ef1ad80-c206-11e9-9a6f-9d5e1e6b832f.png)

![refresh and extend](https://user-images.githubusercontent.com/2958461/63235152-63b66180-c206-11e9-9be4-90cee1159cbb.png)

**Comment by TheDuckCow** [comment]
Minor update, this feature has not been forgotten, but has not been heavily prioritized due to other things that need to be done first before this can be launched effectively:

- New updates for all the block mappings for Vanilla 1.14+, this is a manual step and I am reconciling how to best implement this so it is reasonably future proof and easy to update as things continue to change.
- Resolving existing stability and usability issues
- Making sure I'm braced for a waterfall of issues and support questions, as this inevitably will be a complex feature where things can go wrong in many different ways. I'll need to implement clear way to help the users help me debug for them, to isolate when the arisen issues are 1) From blender, 2) from Mineways or jmc2obj itself (in which case I redirect to those support pages), or 3) MCprep handling of the interfaces.

**Comment by TheDuckCow** [comment]
Another update, aiming to push out the next version (likely naming 3.3.0), after which this feature should be feasible for limited alpha testing. Due to the limited nature, I'm planning to make an opt-in toggle feature and/or a notice to press OK that feature is experimental, so that it's not depended on for production work (at least not without extra precautions).

Also saving this [nbt link](https://github.com/twoolie/NBT) as I leveraged part of this and planning to bundle or extract the needed piece for reading player locations as a starting point.

**Comment by StandingPadAnimations** [comment]
Revisiting the issue (as well as some other old issues). I think we should create an MCprep Hackathon type of event for stuff that's on the back burner, but also too complex to label a "good first contribution"

**Comment by TheDuckCow** [comment]
This task is being superseded by this discussion: https://github.com/Moo-Ack-Productions/MCprep/discussions/709

We could potentially still use this as a parent task though, we'll decide later.
